solver/vertex.go

```diff
@@ -0,0 +1,92 @@
+package solver
+
+import (
+	"sync"
+	"time"
+
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/util/progress"
+	digest "github.com/opencontainers/go-digest"
+	"golang.org/x/net/context"
+)
+
+type Vertex interface {
+	Digest() digest.Digest
+	Sys() interface{}
+	Inputs() []Input
+	Name() string // change this to general metadata
+}
+
+type Input struct {
+	Index  int
+	Vertex Vertex
+}
+
+type input struct {
+	index  int
+	vertex *vertex
+}
+
+type vertex struct {
+	mu           sync.Mutex
+	sys          interface{}
```

Comment:
- when is this set to non-nil?
- type is not `Reference`?
- variable name is hard to understand without comment :sweat_smile: